### PR TITLE
Add deepcopy_internal method for Dict

### PR DESCRIPTION
**Pull request by tkoolen**
Fixes https://github.com/JuliaLang/julia/issues/19921.

### CODE REVIEW

File: test/copy.jl [hunk 1]
```diff
@@ -108,3 +108,25 @@ let x = BigFloat[1:1000;], y, z, v
...
+
+# issue #19921
+type Foo
+    a::String
+end
+
+type Bar
+    foo::Foo
+    fooDict::Dict{Foo, Int64}
+end
```
Comment (ararslan) [inline test/copy.jl]: The type definitions should be able to go in the `@testset`. That would avoid a bit of scope clutter. Probably not a big deal though.

Comment (tkoolen) [inline test/copy.jl]: When I do that, I get
```
Error in testset copy:
Error During Test
  Got an exception of type LoadError outside of a @test
  LoadError: error compiling anonymous: type definition not allowed inside a local scope
  while loading /home/twan/code/julia/test/copy.jl, in expression starting on line 112
ERROR: LoadError: 
Stacktrace:
 [1] (::##33#41)() at /home/twan/code/julia/test/runtests.jl:195
 [2] cd(::##33#41, ::String) at ./file.jl:69
 [3] include_from_node1(::String) at ./loading.jl:532
 [4] include(::String) at ./sysimg.jl:14
 [5] process_options(::Base.JLOptions) at ./client.jl:308
 [6] _start() at ./client.jl:374
while loading /home/twan/code/julia/test/runtests.jl, in expression starting on line 38
```
I'll rename the types to something more specific instead.

File: base/deepcopy.jl [hunk 2]
```diff
@@ -75,3 +75,22 @@ function _deepcopy_array_t(x::ANY, T, stackdict::ObjectIdDict)
...
+
+    dest = similar(x)
+    stackdict[x] = dest
+    for (k, v) in x
+        dest[Base.deepcopy_internal(k, stackdict)] = Base.deepcopy_internal(v, stackdict)
+    end
+    dest
+end
+
+
```
Comment (yuyichao) [inline base/deepcopy.jl]: Minor issue: Too many empty lines at the end....

File: test/copy.jl (hunk 1 above)
Comment (yuyichao) [inline test/copy.jl]: I don't think `@testset` allows toplevel-only expressions. Also types can't be namespaced within the same module.

Comment (ararslan) [inline test/copy.jl]: Oh sorry, I didn't realize you couldn't create types in testsets! Good to know.

File: base/deepcopy.jl (hunk 2 above)
Comment (tkoolen) [inline base/deepcopy.jl]: Fixed.

**Comment by tkoolen** [comment]
Any other issues?

**Pull request merged**